### ISSUE TITLE
App not running on RPi4 (64-bit OS)

### ISSUE THREAD

**Issue by craftbyte** [balenalabs/balena-dash#43]
I wanted to try and see if the project works on a RPi4 that I got my hands on, but was left disappointed. I tried increasing the GPU MEM to 256 MB.

Granted support access for 1 week: https://dashboard.balena-cloud.com/devices/2bd2fcff4d8f39444cf03a94d498115f/actions

**Comment by dansku** [comment]
Hey @craftbyte, thanks for the issue. 
We are aware of the issue, the balaneDash project is currently not working with balenaOS 2.41.0+rev4.
I will keep you posted with updates.
Thank you

**Comment by dansku** [comment]
@craftbyte Actually WPE is currently not working on 64-bit yet. :)

**Comment by floion** [comment]
wpe will need to be recompiled to use the open source gl libraries because the pi4 apparently will not support the legacy closed source gl drivers:

```
No. The firmware support has been removed. Moving to the much more standardised Mesa/DRM/KMS system is the recommended approach, it provides much better support.

SORRY, just been pointed out you are on 3B, I was talking about the 4. Apologies. So what you are trying to do SHOULD work.
```

from https://github.com/raspberrypi/linux/issues/3206#issuecomment-528303004

So this means that we need to have wpe use the `fdo` backend instead of the `rdk` backend here: https://github.com/balena-io-projects/balena-wpe/blob/master/base-image/meta-resin-wpe/conf/samples/local.conf.sample#L35

And also when changing the backend to fdo this would need to be commented out: https://github.com/balena-io-projects/balena-wpe/blob/master/base-image/meta-resin-wpe/conf/samples/local.conf.sample#L36

And then some more changes need to be done to make it display using weston as this new config will use weston.

**Comment by floion** [comment]
Created a balena-wpe issue for this: https://github.com/balena-io-projects/balena-wpe/issues/58

**Comment by JeffersonBledsoe** [comment]
Just wondering how high up the priority list this issue is? Degrades the out-the-box experience quite a bit not having immediate access to just download this repo and push it to the device. It works so well on the Pi3!

**Comment by damianpowell** [comment]
I notice that there’s a branch containing what looks like a functioning RPi4 implementation based on Chromium. Is there any reason that it hasn’t been merged to the main branch yet?

**Comment by lukecurtis93** [comment]
Any update on this?

**Comment by rahul-thakoor** [comment]
the chromium-based branch is work in progress and we are still working out some issues. We would be grateful if you can provide your feedback if you've tested it. Thanks :)

**Comment by lukecurtis93** [comment]
thanks for your reply @rahul-thakoor so yeah right now following the tutorial here on a pi4: https://www.balena.io/blog/make-a-web-frame-with-raspberry-pi-in-30-minutes/ simply just returns the splash screen on the balenaOS screen

**Comment by rahul-thakoor** [comment]
yeah @lukecurtis93 since the master branch uses WPE which is not compatible with pi4. you can check out [this branch](https://github.com/balenalabs/balena-dash/tree/add-x86_64-and-rpi4-support) but be aware it is still WIP

**Comment by lukecurtis93** [comment]
thanks @rahul-thakoor so right now I get the following output on the logs when using that branch:
```
10.03.20 14:39:44 (+1000)  wpe  setting xserver-xorg-legacy/xwrapper/allowed_users from configuration file
10.03.20 14:39:44 (+1000)  wpe  Setting Window Size to default
10.03.20 14:39:44 (+1000)  wpe  2020/03/10 04:39:44 Command is '/home/chromium/launch.sh'
10.03.20 14:39:44 (+1000)  wpe  2020/03/10 04:39:44 Starting listening on port 8080
```

The logo simply displays on the screen and wpe doesn't do anything.

Cheers

**Comment by rahul-thakoor** [comment]
you need to set `WPE_URL` or navigate to `<device ip>:8080` to get the URL management UI

**Comment by lukecurtis93** [comment]
thanks that worked!

**Comment by damianpowell** [comment]
> We would be grateful if you can provide your feedback if you've tested it. Thanks :)

Sure! Well it seems pretty stable in my testing so far, at least the basic functions.

The one are where I'm having a problem is rotating the display. I'm trying to rotate 270 degrees (i.e. left). I have set the `ROTATE_DISPLAY` environment variable and that initially appeared to work, except that the touch panel didn't rotate with the visuals! I am currently trying to rotate using the RPi4's `display_hdmi_rotate` setting in `config.txt` (via the `RESIN_HOST_CONFIG_xxx` device configuration). This appears to be less successful - firstly, the Balena splash screen doesn't appear to change shape; it rotates correctly but maintains a landscape aspect ratio. The Chromium engine in the wpe container doesn't fare well either; it displays a completely blank screen. If I use Tohora to browse to a new page, the balena logo is displayed for a few seconds while Chromium restarts but then we go back to the blank screen.

I'm happy to dig into anything that you think might be helpful, but I should add the caveat that I'm fairly ignorant of the lower levels of both Linux and Raspberry Pis! If I make any progress independently, I'll post an update here.

**Comment by rahul-thakoor** [comment]
closed by #64
